src/proxy/connection_manager.rs

```diff
@@ -0,0 +1,359 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::proxy::error;
+use crate::rbac::Connection;
+use crate::state::DemandProxyState;
+use drain;
+use std::collections::HashMap;
+use std::sync::{atomic::AtomicUsize, Arc};
+use tokio::sync::watch;
+use tokio::sync::RwLock;
+use tracing::{debug, info};
+
+struct ConnectionDrain {
+    tx: drain::Signal,
+    rx: drain::Watch,
+    count: AtomicUsize,
+}
+
+impl ConnectionDrain {
+    fn new() -> Self {
+        let (tx, rx) = drain::channel();
+        let count = AtomicUsize::new(1);
+        ConnectionDrain { tx, rx, count }
+    }
+
+    /// drain drops the internal reference to rx and then signals drain on the tx
+    // always inline, this is for convenience so that we don't forget to drop the rx but there's really no reason it needs to grow the stack
+    #[inline(always)]
+    async fn drain(self) {
+        drop(self.rx); // very important, drain cannont complete if there are outstand rx
+        self.tx.drain().await;
+    }
+}
+
+#[derive(Clone)]
+pub struct ConnectionManager {
+    drains: Arc<RwLock<HashMap<Connection, ConnectionDrain>>>,
+}
+
+impl ConnectionManager {
+    pub fn new() -> Self {
+        ConnectionManager {
+            drains: Arc::new(RwLock::new(HashMap::new())),
+        }
+    }
+
+    // register a connection with the manager and get a channel to receive on
+    pub async fn track(self, c: &Connection) -> drain::Watch {
+        let cd = ConnectionDrain::new();
+        let rx = cd.rx.clone();
+        let mut drains = self.drains.write().await;
+        if let Some(w) = drains.remove(c) {
```

Comment:
nit: del+mutate+insert can be done with code like:

```
map.entry("poneyland")
   .and_modify(|e| { *e += 1 })
   .or_insert(42);
```

Reply:
maybe not with the `return rx` aspect though

Reply:
[modified rust playground example](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=06454ee0114a4d96054c4d32f55dcb5c)

Seems like it could work. TY

Reply:
could very well be a step too far and difficult to grok but:

```rust
    // register a connection with the manager and get a channel to receive on
    pub async fn track(self, c: &Connection) -> drain::Watch {
        let cd = ConnectionDrain::new();
        self.drains
            .write()
            .await
            .entry(c.to_owned())
            .and_modify(|cd| cd.count += 1)
            .or_insert(cd)
            .rx
            .clone()
    }
```